### PR TITLE
Feat/dcat validation rebased

### PR DESCRIPTION
**Pull request by jobulcke**
Reusable dcat validation implementation

### CODE REVIEW

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/DcatCatalogValidator.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package be.vlaanderen.informatievlaanderen.ldes.server.domain.validation.dcat;
+
+import be.vlaanderen.informatievlaanderen.ldes.server.domain.validation.dcat.blanknodevalidators.DcatBlankCatalogNode;
+import be.vlaanderen.informatievlaanderen.ldes.server.domain.validation.dcat.cannotcontainvalidators.CannotContainDatasetValidator;
+import be.vlaanderen.informatievlaanderen.ldes.server.domain.validation.dcat.cannotcontainvalidators.CannotContainServiceValidator;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DcatCatalogValidator extends DcatValidator {
+	DcatCatalogValidator() {
```
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/DcatCatalogValidator.java:10]: suggestion: Unless, there's a good reason to have no access modifier, I'd prefer if you would public access modifier here.

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/blanknodevalidators/DcatBlankCatalogNode.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import org.apache.jena.rdf.model.Model;
+import org.apache.jena.rdf.model.Resource;
+import org.apache.jena.vocabulary.RDF;
+
+import java.util.List;
+
+import static be.vlaanderen.informatievlaanderen.ldes.server.domain.validation.dcat.DcatValidator.DCAT_CATALOG;
+
+public class DcatBlankCatalogNode implements DcatBlankNodeValidator {
```
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/blanknodevalidators/DcatBlankCatalogNode.java:11]: DcatBlankCatalogNode is in terms of naming a bit inconsistent -> DcatBlankCatalogNodeValidator

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/blanknodevalidators/DcatBlankDataServiceNode.java
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/blanknodevalidators/DcatBlankDataServiceNode.java:1]: Naming consistency -> DcatBlankDataServiceNodeValidator

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/blanknodevalidators/DcatBlankDatasetNode.java
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/blanknodevalidators/DcatBlankDatasetNode.java:1]: naming consistency -> ...Validator

**Comment by WLefever-Cegeka** [comment]
Classes `DcatBlankCatalogNode`, `DcatBlankDataServiceNode` and `DcatBlankDatasetNode` seem to share a very similar structure. It might be interesting to extract this into a Strategy or a Template.

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/DcatValidator.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	public static final Property DCAT_DATASET = createProperty(DCAT, "Dataset");
+	public static final Property DCAT_CATALOG = createProperty(DCAT, "Catalog");
+	public static final Property DCAT_SERVES_DATASET = createProperty(DCAT, "servesDataset");
+	public static final Property DCAT_DATASET_PREDICATE = createProperty(DCAT, "dataset");
+	public static final Property DCAT_DATA_SERVICE_PREDICATE = createProperty(DCAT, "service");
+
+	private final DcatBlankNodeValidator blankNodeValidator;
+	private final List<CannotContainValidator> cannotContainValidators;
+
+	protected DcatValidator(DcatBlankNodeValidator blankNodeValidator,
```
Comment (Tomvbe) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/DcatValidator.java:26]: Why do we have two different Validator interfaces? DcatBlankNodeValidator and CannotContainValidator, they have the same method and you do the same with them. Can't we decrease complexity by mergin these two in one ModelValidator?

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/cannotcontainvalidators/CannotContainDatasetValidator.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+			throw new IllegalArgumentException("Model cannot contain a relation to the dataset.");
+		}
+
+		if (dcat.listSubjectsWithProperty(RDF.type, DCAT_DATASET).hasNext()) {
+			throw new IllegalArgumentException("Model cannot contain a dataset.");
+		}
+
+		if (dcat.listSubjectsWithProperty(DCAT_DATASET_PREDICATE).hasNext()) {
+			throw new IllegalArgumentException("Model cannot contain a relation to the dataset.");
+		}
```
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/cannotcontainvalidators/CannotContainDatasetValidator.java:21]: I'm always a bit afraid when I see multiple if statements in a method. Maybe a refactor to decrease the number of If Statements can help. Suggestion: rule engine and do an all match on the rule engine. https://www.baeldung.com/java-replace-if-statements#rule-engine

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/cannotcontainvalidators/CannotContainDatasetValidator.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+				dcat -> !dcat.listSubjectsWithProperty(DCAT_SERVES_DATASET).hasNext(),
+				dcat -> !dcat.listSubjectsWithProperty(RDF.type, DCAT_DATASET).hasNext(),
+				dcat -> !dcat.listSubjectsWithProperty(DCAT_DATASET_PREDICATE).hasNext());
+	}
+
+	@Override
+	public void validate(Model dcat) {
+		boolean isValid = rules.stream()
+				.map(rule -> rule.evaluate(dcat))
+				.reduce(true, (prevResult, result) -> prevResult && result);
```
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/cannotcontainvalidators/CannotContainDatasetValidator.java:25]: allMatch?

File: ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/cannotcontainvalidators/CannotContainServiceValidator.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+		rules = List.of(
+				dcat -> !dcat.listSubjectsWithProperty(RDF.type, DCAT_DATA_SERVICE).hasNext(),
+				dcat -> !dcat.listSubjectsWithProperty(DCAT_DATA_SERVICE_PREDICATE).hasNext());
+	}
+
+	@Override
+	public void validate(Model dcat) {
+		boolean isValid = rules.stream()
+				.map(rule -> rule.evaluate(dcat))
+				.reduce(true, (prevResult, result) -> prevResult && result);
```
Comment (WLefever-Cegeka) [inline ldes-server-domain/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/domain/validation/dcat/cannotcontainvalidators/CannotContainServiceValidator.java:25]: allMatch

**Review by WLefever-Cegeka — APPROVED** [review]

**Pull request merged**